### PR TITLE
Add immtuable CLPForwardIndex creator and related classes

### PR DESCRIPTION
**Pull request by jackluo923**
<h2>Pull Request Description</h2>

<p>This PR introduces the <code>CLPForwardIndexCreatorV2</code>, an improved forward index creator that generates immutable forward indexes from <code>CLPMutableForwardIndexV2</code>. Depending on the configuration and data characteristics, the forward index can either be dictionary-encoded using CLP or raw-bytes-encoded.</p>

<h4>Key Improvements Over <code>CLPForwardIndexCreatorV1</code>:</h4>

<ol>
  <li>
    <strong>Improved Compression Ratio:</strong>
    <ul>
      <li>This version replaces fixed-bit encoding with fixed-byte encoding and incorporates configurable chunk compression for dictionary-encoded IDs.</li>
      <li>In real-world scenarios, where dictionary entries are large, fixed-bit encoding often fails to achieve optimal compression. The combination of fixed-byte encoding and compression significantly enhances the compression ratio.</li>
    </ul>
  </li>

  <li>
    <strong>Upgrade to V5 Writer Version:</strong>
    <ul>
      <li>Uses the V5 forward index writer, which is optimized for multi-value, fixed-width data types (such as longs and ints).</li>
      <li>The V5 writer improves the storage efficiency for dictionary variable IDs and encoded variables, directly benefiting the compression efficiency of <code>CLPForwardIndexCreatorV2</code>.</li>
    </ul>
  </li>

  <li>
    <strong>Reduced Serialization/Deserialization Overhead:</strong>
    <ul>
      <li>In <code>CLPForwardIndexCreatorV1</code>, each row had to be decoded and re-encoded during conversion, introducing significant serdes overhead. CLPForwardIndexCreatorV2 is designed to avoids this. The dictionary entries from mutable forward index are moved over to the immutable forward index. The encoded data themselves are currently decoded and re-encoded for now, but a followup PR containing the serdes optimizations will avoid the encoding and decoding entirely. This follow-up PR will also preserve the support for sorted table and columnar ingestion.</li>
      <li>Additionally, primitive types (<code>byte[]</code>) are now used for forward indexes, reducing the need to box strings into <code>String</code> objects. This leads to better performance, reduced memory usage, and lower garbage collection overhead.</li>
    </ul>
  </li>
</ol>

<p>This PR enhances Pinot’s forward index creation process by providing better compression, performance improvements, and more efficient handling of multi-value data.</p>

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/forward/CLPMutableForwardIndexV2.java [hunk 1]
```diff
@@ -226,11 +238,9 @@ public void setString(int docId, String value) {
    * encoded message by replacing them with empty arrays, as Pinot does not accept null values.
    *
    * @param clpEncodedMessage The {@link EncodedMessage} to append.
-   * @throws IOException if an I/O error occurs during the appending process.
    */
-  public void appendEncodedMessage(@NotNull EncodedMessage clpEncodedMessage)
-      throws IOException {
-    if (_isClpEncoded) {
+  public void appendEncodedMessage(@NotNull EncodedMessage clpEncodedMessage) {
+    if (_isClpEncoded || _forceEnableClpEncoding) {
```
Comment (deemoliu) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/forward/CLPMutableForwardIndexV2.java:243]: what is _forceEnableClpEncoding? can you provide an example?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/converter/stats/MutableNoDictionaryColStatistics.java [hunk 2]
```diff
@@ -125,4 +125,13 @@ public CLPStats getCLPStats() {
     throw new IllegalStateException(
         "CLP stats not available for column: " + _dataSourceMetadata.getFieldSpec().getName());
   }
+
```
Comment (deemoliu) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/converter/stats/MutableNoDictionaryColStatistics.java:128]:
```
public CLPStats getCLPStats() {
    if (_forwardIndex instanceof CLPMutableForwardIndex) {
      return ((CLPMutableForwardIndex) _forwardIndex).getCLPStats();
    } else if (_forwardIndex instanceof CLPMutableForwardIndexV2) {
      return ((CLPMutableForwardIndexV2) _forwardIndex).getCLPStats();
    }
```
Saw similar code in getCLPStats() function, can you please let me know in which scenarios these functions are being called.

Comment (jackluo923) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/converter/stats/MutableNoDictionaryColStatistics.java:128]: ClpMutableForwardIndexV2 index -> ClpForwardIndexCreator (V1): use `getCLPStats()`
ClpMutableForwardIndexV2 index -> ClpForwardIndexCreatorV2: use `getCLPV2Stats()`

`getCLPStats()` method is basically for compatibility with the initial version of immutable forward index.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/CLPForwardIndexCreatorV2.java [hunk 3]
```diff
@@ -164,18 +180,29 @@ public CLPForwardIndexCreatorV2(File baseIndexDir, ColumnStatistics columnStatis
...
+      ChunkCompressionType chunkCompressionType, boolean forceRawEncoding)
       throws IOException {
     _chunkCompressionType = chunkCompressionType;
 
     // Pick up metadata from mutable forward index and use it to initialize the immutable forward index
     _column = clpMutableForwardIndex.getColumnName();
     _numDoc = clpMutableForwardIndex.getNumDoc();
-    _isClpEncoded = clpMutableForwardIndex.isClpEncoded();
+
+    _intermediateFilesDir =
```
Comment (deemoliu) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/CLPForwardIndexCreatorV2.java:196]: will we cleanup the intermediateFilesDir ?

Comment (jackluo923) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/CLPForwardIndexCreatorV2.java:196]: Yes, it is cleaned up in the `close()` method
```
 @Override
  public void close()
      throws IOException {
    // Delete all temp files
    FileUtils.deleteDirectory(_intermediateFilesDir);   <-  delete directory
    _dataFile.close();
  }
```

**Review by deemoliu — APPROVED** [review]

**Pull request merged**